File: android/src/main/java/expo/modules/beautycamera/BeautyCameraNativeModule.kt
package expo.modules.beautycamera

import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class BeautyCameraNativeModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("BeautyCameraNative")

    View(BeautyCameraNativeView::class) {
      Events("onReady", "onError", "onFaceState")

      Prop("active") { view, active: Boolean ->
        view.active = active
      }

      Prop("facing") { view, facing: String ->
        view.facing = facing
      }

      Prop("beautyEnabled") { view, enabled: Boolean ->
        view.beautyEnabled = enabled
      }

      Prop("smoothingStrength") { view, strength: Double ->
        view.smoothingStrength = strength.toFloat().coerceIn(0f, 0.3f)
      }

      Prop("slimFaceStrength") { view, strength: Double ->
        view.slimFaceStrength = strength.toFloat().coerceIn(0f, 1f)
      }

      Prop("enlargeEyesStrength") { view, strength: Double ->
        view.enlargeEyesStrength = strength.toFloat().coerceIn(0f, 1f)
      }

      Prop("noseSlimStrength") { view, strength: Double ->
        view.noseSlimStrength = strength.toFloat().coerceIn(0f, 1f)
      }

      Prop("enableTorch") { view, enabled: Boolean ->
        view.enableTorch = enabled
      }

      OnViewDestroys { view ->
        view.dispose()
      }

      AsyncFunction("capture") { view: BeautyCameraNativeView, options: Map<String, Double>, promise: Promise ->
        view.capture(
          maxWidth = (options["maxWidth"] ?: 1024.0).toInt(),
          quality = (options["quality"] ?: 0.8).coerceIn(0.1, 1.0),
          promise = promise
        )
      }
    }
  }
}

File: ios/BeautyCameraNativeModule.swift
import ExpoModulesCore

public final class BeautyCameraNativeModule: Module {
  public func definition() -> ModuleDefinition {
    Name("BeautyCameraNative")

    View(BeautyCameraNativeView.self) {
      Events("onReady", "onError", "onFaceState")

      Prop("active") { (view, active: Bool) in
        view.active = active
      }

      Prop("facing") { (view, facing: String) in
        view.facing = facing
      }

      Prop("beautyEnabled") { (view, enabled: Bool) in
        view.beautyEnabled = enabled
      }

      Prop("smoothingStrength") { (view, strength: Double) in
        view.smoothingStrength = Float(max(0, min(0.3, strength)))
      }

      Prop("slimFaceStrength") { (view, strength: Double) in
        view.slimFaceStrength = Float(max(0, min(1, strength)))
      }

      Prop("enlargeEyesStrength") { (view, strength: Double) in
        view.enlargeEyesStrength = Float(max(0, min(1, strength)))
      }

      Prop("noseSlimStrength") { (view, strength: Double) in
        view.noseSlimStrength = Float(max(0, min(1, strength)))
      }

      Prop("enableTorch") { (view, enabled: Bool) in
        view.torchEnabled = enabled
      }

      AsyncFunction("capture") { (
        view: BeautyCameraNativeView,
        options: [String: Double]
      ) in
        let maxWidth = Int(options["maxWidth"] ?? 1024)
        let quality = options["quality"] ?? 0.8
        return try await view.capture(
          maxWidth: maxWidth,
          quality: quality
        )
      }
    }
  }
}
